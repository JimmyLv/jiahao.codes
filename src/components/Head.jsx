import React from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';
import { site as sitePropType } from '../proptypes';

// TODO: find a better image
const siteCoverImage = require('../../static/favicon.ico');

const buildStructuredData = ({ siteUrl, siteTitle }) =>
  JSON.stringify([
    {
      '@context': 'http://schema.org',
      '@type': 'WebSite',
      url: siteUrl,
      name: siteTitle,
      alternateName: siteTitle,
    },
    // TODO: more things for article https://developers.google.com/search/docs/data-types/article#non-amp
  ]);

function HeadComponent({
  description,
  image,
  structuredData,
  url,
  title,
  isPost,
  facebookAppId,
  twitterUser,
}) {
  return (
    <Helmet>
      <title>{title}</title>

      <meta name="description" content={description} />
      <meta name="image" content={image} />

      <script type="application/ld+json">{structuredData}</script>

      <meta property="og:url" content={url} />
      {isPost ? <meta property="og:type" content="article" /> : null}
      <meta property="og:title" content={title} />
      <meta property="og:description" content={description} />
      <meta property="og:image" content={image} />
      <meta property="fb:app_id" content={facebookAppId} />

      <meta name="twitter:card" content="summary_large_image" />
      <meta name="twitter:creator" content={twitterUser} />
      <meta name="twitter:title" content={title} />
      <meta name="twitter:description" content={description} />
      <meta name="twitter:image" content={image} />
    </Helmet>
  );
}

HeadComponent.propTypes = {
  description: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  structuredData: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  isPost: PropTypes.bool.isRequired,
  facebookAppId: PropTypes.string.isRequired,
  twitterUser: PropTypes.string.isRequired,
};

export default function Head({
  image,
  path,
  excerpt,
  title,
  site: {
    siteMetadata: {
      siteUrl,
      title: siteTitle,
      description: siteDescription,
      facebookAppId,
      twitterUser,
    },
  },
}) {
  return (
    <HeadComponent
      {...{
        description: excerpt || siteDescription,
        image: image ? siteUrl + image : siteCoverImage,
        url: siteUrl + path,
        title: title || siteTitle,
        isPost: !!path,
        structuredData: buildStructuredData({ siteUrl, siteTitle }),
        facebookAppId,
        twitterUser,
      }}
    />
  );
}

Head.propTypes = {
  excerpt: PropTypes.string,
  image: PropTypes.string,
  title: PropTypes.string,
  path: PropTypes.string,
  site: sitePropType.isRequired,
};

Head.defaultProps = {
  excerpt: '',
  image: '',
  title: '',
  path: '',
};
