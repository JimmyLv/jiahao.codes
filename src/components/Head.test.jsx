import React from 'react';
import { shallow } from 'enzyme';
import Head from './Head';

test('rendering', () => {
  const props = {
    excerpt: 'excerpt',
    image: 'image',
    title: 'title',
    path: 'path',
    site: {
      siteMetadata: {
        siteUrl: 'siteUrl',
        title: 'title',
        author: 'author',
        description: 'description',
        facebookAppId: 'facebookAppId',
        twitterUser: 'twitterUser',
        social: {
          githubUrl: 'githubUrl',
          keybaseUrl: 'keybaseUrl',
          linkedInUrl: 'linkedInUrl',
          twitterUrl: 'twitterUrl',
        },
      },
    },
  };
  const wrapper = shallow(<Head {...props} />);
  expect(wrapper).toMatchSnapshot();
});
