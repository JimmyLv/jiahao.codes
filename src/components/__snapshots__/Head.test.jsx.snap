// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering 1`] = `
<HeadComponent
  description="excerpt"
  facebookAppId="facebookAppId"
  image="siteUrlimage"
  isPost={true}
  structuredData="[{\\"@context\\":\\"http://schema.org\\",\\"@type\\":\\"WebSite\\",\\"url\\":\\"siteUrl\\",\\"name\\":\\"title\\",\\"alternateName\\":\\"title\\"}]"
  title="title"
  twitterUser="twitterUser"
  url="siteUrlpath"
/>
`;
